src/librbd/api/Mirror.cc

```diff
@@ -3144,96 +3148,42 @@ int Mirror<I>::group_promote(IoCtx& group_ioctx, const char *group_name,
-        std::vector<cls::rbd::GroupImageSpec> inserted_images; // to current membership
-        if (!delta_images.empty()) {
-          for (auto &s : delta_images) {
-            if (std::find(rollback_images.begin(), rollback_images.end(), s)
-                != rollback_images.end()) {
-              removed_images.push_back(s);
-            } else {
-              inserted_images.push_back(s);
-            }
-          }
-        }
-        if (!removed_images.empty()) {
-          // insert back to membership
-          for (auto &s : removed_images) {
-            IoCtx image_ioctx;
-            r = librbd::util::create_ioctx(group_ioctx, "image", s.pool_id, {},
-                                           &image_ioctx);
-            if (r < 0) {
-              return r;
-            }
-
-            r = Group<I>::image_add(group_ioctx, group_name,
-                                    image_ioctx, s.image_id.c_str());
-            if (r < 0) {
-              lderr(cct) << "error inserting image to group: " << group_name
-                         << " :" << cpp_strerror(r) << dendl;
-              return r;
-            }
-          }
-        }
-
-        if(!inserted_images.empty()) {
-          // remove from the membershiip
-          for (auto &s : inserted_images) {
-            IoCtx image_ioctx;
-            r = librbd::util::create_ioctx(group_ioctx, "image", s.pool_id, {},
-                                           &image_ioctx);
-            if (r < 0) {
-              return r;
-            }
-            r = Group<I>::image_remove(group_ioctx, group_name,
-                                       image_ioctx, s.image_id.c_str());
-            if (r < 0 && r != -ENOENT) {
-              lderr(cct) << "error removing image from a group: " << group_name
-                         << " :" << cpp_strerror(r) << dendl;
-              return r;
-            }
-          }
-        }
-      } // Fixing membership done.
+    if (snap == snaps.rend()) {
+      lderr(cct) << "failed to rollback, no mirror group snapshot available"
+        << dendl;
+      return -EINVAL;
+    }
+    // Check for group membership match
+    std::vector<cls::rbd::GroupImageSpec> rollback_images;
+    for (auto& it : snap->snaps) {
+      rollback_images.emplace_back(it.image_id, it.pool);
+    }
 
-      librbd::NoOpProgressContext prog_ctx;
-      r = Group<I>::snap_rollback(group_ioctx,
-                                  group_name, snap->name.c_str(), prog_ctx);
-      if (r < 0) {
-        lderr(cct) << "failed to rollback to group snapshot: " << snap->name
-                   << " :" << cpp_strerror(r) << dendl;
-        return r;
-      }
-      ldout(cct, 5) << "successfully rollbacked to group snapshot: "
-                    << snap->name << dendl;
-      // Rollback to last good snapshot done
-    } else {
-      lderr(cct) << "failed to rollback, no initial group snapshot available"
+    if (rollback_images != current_images) {
```

Comment:
Does this even need to be checked here seeing that dynamic groups are not supported?

If it does, this could cause the following issue for dynamic groups:

1. Group membership changes on the primary - new images are added
2. The secondary is notified somehow and restarts the GR
3. The bootstrap determines that the membership has changed and creates the IRs for the new images
4. The IRs start up first and the IR Bootstrap routine will create the new image and add it to the group
5. The GR starts up again - at this point it has not created the group new mirror snapshot
6. A force promote is called on the secondary. There is an older complete mirror group snap but its membership will not match that of the group now.